pkg/controllers/jobset_controller.go

```diff
@@ -603,7 +607,174 @@ func executeSuccessPolicy(ctx context.Context, js *jobset.JobSet, ownedJobs *chi
+		return "", false
+	}
+
+	replicatedJobNameKey := job.Labels[jobset.ReplicatedJobNameKey]
+	replicatedJobNameIsUnset := replicatedJobNameKey == ""
+	return replicatedJobNameKey, !replicatedJobNameIsUnset
+}
+
+// If the rule is not nil, then the functions returns:
+// - true if the rule is applicable to the jobFailureReason;
+// - false other.
+// If the rule is nil, we return true.
+func ruleIsApplicable(rule *jobset.FailurePolicyRule, failedJob *batchv1.Job, jobFailureReason string) bool {
+	if rule == nil {
+		return true
+	}
+
+	ruleAppliesToJobFailureReason := len(rule.OnJobFailureReasons) == 0 || slices.Contains(rule.OnJobFailureReasons, jobFailureReason)
+	if !ruleAppliesToJobFailureReason {
+		return false
+	}
+
+	parentReplicatedJob, exists := parentReplicatedJobName(failedJob)
+	if !exists {
+		// If we cannot find the parent ReplicatedJob, we assume the rule does not apply.
+		return false
+	}
+
+	ruleAppliesToParentReplicatedJob := len(rule.TargetReplicatedJobs) == 0 || slices.Contains(rule.TargetReplicatedJobs, parentReplicatedJob)
+	return ruleAppliesToParentReplicatedJob
+}
+
+// The type failurePolicyActionApplier applies a FailurePolicyAction and returns nil if the FailurePolicyAction was successfully applied.
+// The function returns an error otherwise.
+type failurePolicyActionApplier = func(ctx context.Context, js *jobset.JobSet, matchingFailedJob *batchv1.Job, updateStatusOpts *statusUpdateOpts) error
+
+// -------------------------Helper Functions for Each FailurePolicyActions Start------------------------
+
+// failJobSetActionApplier applies the FailJobSet FailurePolicyAction
+var failJobSetActionApplier failurePolicyActionApplier = func(ctx context.Context, js *jobset.JobSet, matchingFailedJob *batchv1.Job, updateStatusOpts *statusUpdateOpts) error {
+	setJobSetFailedCondition(ctx, js, constants.ReachedMaxRestartsReason, messageWithFirstFailedJob(constants.ReachedMaxRestartsMessage, matchingFailedJob.Name), updateStatusOpts)
+	return nil
+}
+
+// restartJobSetActionApplier applies the RestartJobSet FailurePolicyAction
+var restartJobSetActionApplier failurePolicyActionApplier = func(ctx context.Context, js *jobset.JobSet, matchingFailedJob *batchv1.Job, updateStatusOpts *statusUpdateOpts) error {
+	if js.Status.RestartsCountTowardsMax >= js.Spec.FailurePolicy.MaxRestarts {
+		setJobSetFailedCondition(ctx, js, constants.ReachedMaxRestartsReason, messageWithFirstFailedJob(constants.ReachedMaxRestartsMessage, matchingFailedJob.Name), updateStatusOpts)
+		return nil
+	}
+
+	shouldCountTowardsMax := true
+	failurePolicyRecreateAll(ctx, js, shouldCountTowardsMax, updateStatusOpts)
+	return nil
+}
+
+// restartJobSetAndIgnoreMaxRestartsActionApplier applies the RestartJobSetAndIgnoreMaxRestarts FailurePolicyAction
+var restartJobSetAndIgnoreMaxRestartsActionApplier failurePolicyActionApplier = func(ctx context.Context, js *jobset.JobSet, matchingFailedJob *batchv1.Job, updateStatusOpts *statusUpdateOpts) error {
```

Comment:
I think we should emit an event when a failure policy rule is applied, so there is better observability for the user and they can easily understand why the JobSet was restarted/marked failed/restarted without counting against max restarts etc. Also will be helpful for debugging.

Reply:
This is done. I modified `failurePolicyRecreateAll` so that the event can be passed in as an argument and wrote messages for the two restart action types to make it clear when the restart counts against the maximum number of allowed restarts.

Please resolve if this is what you had in mind.

Reply:
Right idea so far, one thing we need to change is the event messages should be passed as an [eventMessage](https://github.com/kubernetes-sigs/jobset/blob/454842ab57af7b739eccbd2a2dec36186a4f7652/pkg/controllers/jobset_controller.go#L81) rather than `eventReason`. Reason should be a short machine interpretable string like `"JobSetFooBar`" (by convention we use CamelCase, see Job failure [reasons](https://github.com/kubernetes/kubernetes/blob/201e6262c4158db23b8887a9c31e326d34d989d4/staging/src/k8s.io/api/batch/v1/types.go#L632)), and the message is an English sentence geared toward human readers. 

See constants.go for examples. Let's set the reasons as constants in constants.go as well like the other event reasons, and any messages which are constant - dynamically generated messages which interpolate some variable can remain as is in fmt.Sprintf or a helper.

Reply:
> Reason should be a short machine interpretable string like "`JobSetFoobar`"

If I understand correctly, this is not true for the current implementation of  `failurePolicyRecreateAll` as done [here](https://github.com/kubernetes-sigs/jobset/blob/1d1c2f3fd33293422d801f590f81cb496f609de8/pkg/controllers/jobset_controller.go#L638). Is it documented anywhere that the event reason should follow that formatting?

Reply:
Yeah the linked example should not be that way, that is a mistake, that reason should be a message. We can fix it here or in a follow up PR. You can search the string `"eventRecorder.Event"` in the kubernetes codebase to see examples of event reasons and messages and their formatting conventions.

Reply:
Sounds good. I updated the code with different reasons and messages for each failure policy action type.

I also added comments to the definition of `eventParams` in `pkg/controllers/jobset_controller.go`. Please take a look and let me know if the definitions I added are not accurate.

Reply:
Looks good to me